libretro-common/queues/generic_queue.c

```diff
@@ -0,0 +1,356 @@
+{
+   if (!queue || !queue->first_item)
+   {
+      return NULL;
+   }
+
+   return queue->first_item->value;
+}
+
+void generic_queue_shift(generic_queue_t *queue, void *value)
+{
+   struct generic_queue_item_t *new_item;
+
+   if (!queue)
+   {
+      return;
+   }
+
+   new_item = (struct generic_queue_item_t *)calloc(1, sizeof(struct generic_queue_item_t));
+   new_item->value = value;
+   new_item->previous = NULL;
+   new_item->next = queue->first_item;
+
+   queue->first_item = new_item;
+   queue->length++;
+
+   if (!queue->last_item)
+   {
+      queue->last_item = new_item;
+   } else
+   {
+      new_item->next->previous = new_item;
+   }
+}
+
+void *generic_queue_unshift(generic_queue_t *queue)
+{
+   void *value;
+
+   if (!queue || !queue->first_item)
+   {
+      return NULL;
+   }
+
+   struct generic_queue_item_t *item;
+
+   item = queue->first_item;
+   queue->first_item = queue->first_item->next;
+   queue->length--;
+
+   if (queue->length == 0)
+   {
+      queue->last_item = NULL;
+   }
+
+   value = item->value;
+   free(item);
+
+   return value;
+}
+
+size_t generic_queue_length(generic_queue_t *queue)
+{
+   if (queue)
+   {
+      return queue->length;
+   } else
+   {
+      return 0;
+   }
+}
+
+void *generic_queue_remove(generic_queue_t *queue, void *value)
+{
+   struct generic_queue_item_t *item;
+
+   if (!queue)
+   {
+      return NULL;
+   }
+
+   for (item = queue->first_item; item; item = item->next)
+   {
+      if (item->value == value)
+      {
+         void *value;
+
+         if (item->previous)
+         {
+            item->previous->next = item->next;
+         } else
+         {
+            queue->first_item = item->next;
+         }
+
+         if (item->next)
+         {
+            item->next->previous = item->previous;
+         } else
+         {
+            queue->last_item = item->previous;
+         }
+
+         value = item->value;
+         free(item);
+         queue->length--;
+
+         return value;
+      }
+   }
+
+   return NULL;
+}
+
+generic_queue_iterator_t *generic_queue_iterator(generic_queue_t *queue, bool forward)
+{
+   if (queue && queue->first_item)
+   {
+      generic_queue_iterator_t *iterator;
+
+      iterator = (generic_queue_iterator_t *)malloc(sizeof(generic_queue_iterator_t));
+      iterator->queue = queue;
+      iterator->item = forward ? queue->first_item : queue->last_item;
+      iterator->forward = forward;
+
+      return iterator;
+   } else
+   {
+      return NULL;
+   }
+}
+
+generic_queue_iterator_t *generic_queue_iterator_next(generic_queue_iterator_t *iterator)
+{
+   if (iterator)
+   {
+      struct generic_queue_item_t *item;
+
+      item = iterator->forward ? iterator->item->next : iterator->item->previous;
+      if (item)
+      {
+         iterator->item = item;
+         return iterator;
+      } else
+      {
+         free(iterator);
+         return NULL;
+      }
+   } else
+   {
```

Comment:
In this case, it's best to get rid of the 'else' block and instead just have 'return NULL' as the final line.

Reply:
Updated.